File: WebFt30/src/app/core/grpc/services/gis.service.ts
import { inject, Injectable } from '@angular/core';
import { GrpcWebFetchTransport } from '@protobuf-ts/grpcweb-transport';
import { AuthInterceptor } from '../auth.interceptor';
import { GrpcUtils } from '../grpc.utils';
import * as grpc from 'src/grpc-generated';
import { Observable } from 'rxjs';
import { FtEnumsMapping } from '../../store/mapping/ft-enums-mapping';

@Injectable({
  providedIn: 'root'
})
export class GisService {
  private authInterceptor = inject(AuthInterceptor);
  private client: grpc.GisClient;

  constructor() {
    const interceptors = [this.authInterceptor.toRpcInterceptor()];
    const options = GrpcUtils.getGrpcOptions(GrpcUtils.getApiUrl(), interceptors);
    const transport = new GrpcWebFetchTransport(options);
    this.client = new grpc.GisClient(transport);
  }

  getAllGeoData(): Observable<grpc.GetAllGeoDataResponse> {
    const request: grpc.GetAllGeoDataRequest = {};
    return GrpcUtils.unaryToObservable(this.client.getAllGeoData.bind(this.client), request, {});
  }

  getFiberInfo(fiberId: string): Observable<grpc.GetFiberInfoResponse> {
    const request: grpc.GetFiberInfoRequest = { fiberId };
    return GrpcUtils.unaryToObservable(this.client.getFiberInfo.bind(this.client), request, {});
  }

  getLandmarks(traceId: string): Observable<grpc.GetLandmarksResponse> {
    const request: grpc.GetLandmarksRequest = { traceId };
    return GrpcUtils.unaryToObservable(this.client.getLandmarks.bind(this.client), request, {});
  }

  getLandmarksModel(landmarksModelId: string): Observable<grpc.GetLandmarksModelResponse> {
    const request: grpc.GetLandmarksModelRequest = { landmarksModelId };
    return GrpcUtils.unaryToObservable(
      this.client.getLandmarksModel.bind(this.client),
      request,
      {}
    );
  }

  createLandmarksModel(
    landmarksModelId: string,
    traceId: string,
    gpsInputMode: string
  ): Observable<grpc.CreateLandmarksModelResponse> {
    const request: grpc.CreateLandmarksModelRequest = {
      landmarksModelId,
      traceId,
      gpsInputMode: FtEnumsMapping.toGrpcGpsInputMode(gpsInputMode)
    };
    return GrpcUtils.unaryToObservable(
      this.client.createLandmarksModel.bind(this.client),
      request,
      {}
    );
  }
}

File: Backend/src/Fibertest30.Api/Services/GisService.cs
using Grpc.Core;
using MediatR;

namespace Fibertest30.Api;

public class GisService(ISender mediator) : Gis.GisBase
{
    public override async Task<GetAllGeoDataResponse> GetAllGeoData(GetAllGeoDataRequest request, ServerCallContext context)
    {
        var allGeoData = await mediator.Send(new GetAllGeoDataQuery(), context.CancellationToken);
        return new GetAllGeoDataResponse() { Data = allGeoData.ToProto() };
    }

    public override async Task<GetFiberInfoResponse> GetFiberInfo(GetFiberInfoRequest request,
        ServerCallContext context)
    {
        var fiberInfo =
            await mediator.Send(new GetFiberInfoQuery(Guid.Parse(request.FiberId)), context.CancellationToken);
        return new GetFiberInfoResponse() { FiberInfo = fiberInfo.ToProto() };
    }

    public override async Task<GetLandmarksResponse> GetLandmarks(GetLandmarksRequest request,
        ServerCallContext context)
    {
        var landmarks =
            await mediator.Send(new GetLandmarksQuery(Guid.Parse(request.TraceId)), context.CancellationToken);
        return new GetLandmarksResponse() { TraceId = request.TraceId, Landmarks = { landmarks.Select(l=>l.ToProto()) } };
    }

    public override async Task<GetLandmarksModelResponse> GetLandmarksModel(GetLandmarksModelRequest request, ServerCallContext context)
    {
        var landmarksModel =
            await mediator.Send(new GetLandmarksModelQuery(Guid.Parse(request.LandmarksModelId)),
                context.CancellationToken);

        return new GetLandmarksModelResponse() { LandmarksModel = landmarksModel.ToProto() };
    }

    public override async Task<CreateLandmarksModelResponse> CreateLandmarksModel(CreateLandmarksModelRequest request,
        ServerCallContext context)
    {
            await mediator.Send(
                new CreateLandmarksModelCommand(Guid.Parse(request.LandmarksModelId),
                    Guid.Parse(request.TraceId), request.GpsInputMode.FromProto()), context.CancellationToken);

        return new CreateLandmarksModelResponse() { };
    }
}
